internal/ethapi/api.go

```diff
@@ -1232,21 +1232,25 @@ func DoEstimateGas(ctx context.Context, b Backend, args TransactionArgs, blockNr
 		if err != nil {
 			return 0, err
 		}
-		if metaTxParams != nil {
-			sponsorAmount, _ := types.CalculateSponsorPercentAmount(metaTxParams, feeCap)
-			sponsorBalance := state.GetBalance(metaTxParams.GasFeeSponsor)
-			if sponsorBalance.Cmp(sponsorAmount) <= 0 {
-				return 0, types.ErrSponsorBalanceNotEnough
-			}
-			balance = new(big.Int).Add(balance, sponsorAmount)
-		}
+
 		available := new(big.Int).Set(balance)
 		if args.Value != nil {
-			if args.Value.ToInt().Cmp(available) >= 0 {
+			if args.Value.ToInt().Cmp(available) > 0 {
 				return 0, core.ErrInsufficientFundsForTransfer
 			}
 			available.Sub(available, args.Value.ToInt())
 		}
+
+		if metaTxParams != nil {
+			sponsorAmount, _ := types.CalculateSponsorPercentAmount(metaTxParams, new(big.Int).Mul(feeCap, new(big.Int).SetUint64(hi)))
+			sponsorBalance := state.GetBalance(metaTxParams.GasFeeSponsor)
+			if sponsorAmount.Cmp(sponsorBalance) < 0 {
+				available.Add(available, sponsorAmount)
+			} else {
+				available.Add(available, sponsorBalance)
```

Comment:
sponser & user's balance are added to `available`, but not all their balance are used to afford gasFee. 
- in a case, sponser's balance is used to afford, and user's balance should not be involved.

Reply:
Because of we can not get the correct gasLimit before EstimateGas，we can only calculate `sponsorAmount` and `selfAmount` by kind of blockGasLimit `feeCap * hi`. The results of two `amount` may be much larger than balance. Such judgments may be inappropriate.